// @antv/l7-layers/es/core/TextureService.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import { TYPES } from '@antv/l7-core';
import { generateColorRamp } from '@antv/l7-utils';

var TextureService = /*#__PURE__*/function () {
  function TextureService(layer) {
    _classCallCheck(this, TextureService);

    this.layer = layer;
    var container = this.layer.getContainer();
    this.rendererService = container.get(TYPES.IRendererService);
  }

  _createClass(TextureService, [{
    key: "getColorTexture",
    value: function getColorTexture(colorRamp) {
      // TODO 支持传入图片
      var currentkey = this.getTextureKey(colorRamp);

      if (this.key === currentkey) {
        return this.colorTexture;
      } else {
        this.createColorTexture(colorRamp);
      }

      this.key = currentkey;
      return this.colorTexture;
    }
  }, {
    key: "createColorTexture",
    value: function createColorTexture(colorRamp) {
      var createTexture2D = this.rendererService.createTexture2D;
      var imageData = generateColorRamp(colorRamp);
      var texture = createTexture2D({
        data: imageData.data,
        width: imageData.width,
        height: imageData.height,
        flipY: false
      });
      this.colorTexture = texture;
      return texture;
    }
  }, {
    key: "setColorTexture",
    value: function setColorTexture(texture, colorRamp) {
      this.key = this.getTextureKey(colorRamp);
      this.colorTexture = texture;
    }
  }, {
    key: "destroy",
    value: function destroy() {
      var _this$colorTexture;

      (_this$colorTexture = this.colorTexture) === null || _this$colorTexture === void 0 ? void 0 : _this$colorTexture.destroy();
    }
  }, {
    key: "getTextureKey",
    value: function getTextureKey(colorRamp) {
      return "".concat(colorRamp.colors.join('_'), "_").concat(colorRamp.positions.join('_'));
    }
  }]);

  return TextureService;
}();

export { TextureService as default };
